### ISSUE TITLE
indexer_update_all_views invalidates all cache after any inventory stock change

### ISSUE THREAD

**Issue by CptCharlesG** [magento/magento2#36813]
### Preconditions and environment

- Magento version: 2.4.5-p1
- Both on FPC and Varnish
- Indexers update on schedule (cron)
- Inventory with only `default` source and stock
- 2 websites, 3 stores (only 2 stores enabled)
- Flat category: Yes
- Flat products: No
- Decrease Stock When Order is Placed: Yes
- Set Items' Status to be In Stock When Order is Cancelled: Yes
- Display Out of Stock Products: Yes
- Only X left Threshold: 0
- Display Products Availability in Stock on Storefront: Yes
- Synchronize with Catalog: Yes
- Manage Stock: Yes
- Manage Stock: No Backorers
- Out-of-Stock Threshold: 0
- Automatically Return Credit Memo Item to Stock: No

If that counts, this system was upgraded from Magento 1.9.4 to 2.4.3 and later every version.

After stock changes, the `indexer_update_all_views `invalidates all cache with "invalidateInfo":{"tags":[],"mode":"all"}}.

Issue appears when creating a shipment or updating the inventory in a module with the following code:
```
$sourceItem = $this->sourceItemFactory->create();
$sourceItem->setSourceCode('default');
$sourceItem->setSku($sku);
$sourceItem->setQuantity($qtyToSet);
$sourceItem->setStatus(($qty > 0 ? 1 : 0));
$this->sourceItemsSaveInterface->execute([$sourceItem]);
```

The invalidation happens the next cron run inside `indexer_update_all_views `.

Possibly they have the same issue in the new comments (since 2022 Q3) on an already closed bug report:
[https://github.com/magento/magento2/issues/30128#issuecomment-1275090668](https://github.com/magento/magento2/issues/30128#issuecomment-1275090668)

Due to it happens for others too, I think it is a core issue, but possibly we have the same 3rd party modules installed.
I tried to check all of these modules for any cache related code, but could not find anything that could cause.

3rd party modules:
```
Amasty_Base
Amasty_CronScheduleList
Amasty_Finder
Amasty_Geoip
Amasty_GdprCookie
Amasty_GdprLite
Amasty_GdprFrontendUi
Amasty_Mage245Fix
Amasty_Mage24Fix
Amasty_SeoSingleUrl
BrunoCanada_HrefLang
Chessio_Matomo
Codazon_AjaxLayeredNavPro
Codazon_Core
Codazon_GoogleAmpManager
Codazon_Lookbookpro
Codazon_MegaMenu
Codazon_ProductFilter
Codazon_ProductLabel
Codazon_QuickShop
Codazon_SalesPro
Codazon_ShippingCostCalculator
Codazon_Shopbybrandpro
Codazon_ShoppingCartPro
Codazon_ThemeLayoutPro
Dan0sz_ResourceHints
Flurrybox_Core
Flurrybox_EnhancedPrivacy
Genaker_ValidateAttributeLength
GhoSter_OutOfStockAtLast
Iazel_RegenProductUrl
Lillik_PriceDecimal
M2Boilerplate_CriticalCss
MageSuite_Magepack
MageWorx_SearchSuiteAutocomplete
Magefan_Community
Magefan_Blog
Magefan_BlogGraphQl
Magefan_WysiwygAdvanced
Magefox_GoogleShopping
Magenest_CacheWarmer
Yireo_LinkPreload
Yireo_NextGenImages
Yireo_Webp2
```

### Steps to reproduce

1. Create a shipment for an order / update the stock programmatically
2. Wait for the next cron run
3. Check page headers/waiting time
4. Verify the issue in debug.log

### Expected result

Invalidate cache only for the changed products and their categories.

### Actual result

Something issues cache invalidation {"tags":[],"mode":"all"}} 4 times significantly earlier (2 sec!) before proceeding to product and category invalidation.

Then the cache misses, and the page takes 2+ seconds to respond, that leads to bad page speed ratings at google and losing customers. 

### Additional information

FPC log:

```
[2023-02-07T09:42:06.090008+00:00] main.INFO: Cron Job magenest_cachewarmer_warm is successfully finished. Statistics: {"sum":0.0032229423522949,"count":1,"realmem":0,"emalloc":144304,"realmem_start":206045184,"emalloc_start":191280120} [] []
[2023-02-07T09:42:06.196229+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"}} []
[2023-02-07T09:42:06.236829+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"}} []
[2023-02-07T09:42:06.237366+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"}} []
[2023-02-07T09:42:06.237529+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"}} []
[2023-02-07T09:42:08.182239+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["cat_p_7913","FPC"],"mode":"matchingTag"}} []
[2023-02-07T09:42:08.231403+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["cat_p_7728","FPC"],"mode":"matchingTag"}} []
[2023-02-07T09:42:08.278293+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["cat_c_p_7","FPC"],"mode":"matchingTag"}} []
[2023-02-07T09:42:08.328399+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["cat_c_2","FPC"],"mode":"matchingTag"}} []
[2023-02-07T09:42:08.379839+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["cat_c_7","FPC"],"mode":"matchingTag"}} []
[2023-02-07T09:42:08.381348+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["cat_p_7913","cat_p_7728","cat_c_p_7","cat_c_2","cat_c_7"],"mode":"matchingAnyTag"}} []
[2023-02-07T09:42:08.381696+00:00] main.INFO: Cron Job indexer_update_all_views is successfully finished. Statistics: {"sum":2.4353058338165,"count":1,"realmem":20971520,"emalloc":27873416,"realmem_start":206045184,"emalloc_start":191825056} [] []
```

Varnish log:

```
[2023-02-07T10:44:05.332559+00:00] main.INFO: Cron Job indexer_update_all_views is run [] []
[2023-02-07T10:44:06.129578+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"}} []
[2023-02-07T10:44:06.196482+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"}} []
[2023-02-07T10:44:06.196830+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"}} []
[2023-02-07T10:44:06.197018+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":[],"mode":"all"}} []
[2023-02-07T10:44:07.557597+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"servers":[{"Laminas\\Uri\\Uri":"http://127.0.0.1:6081/"}],"formattedTagsChunk":"((^|,)cat_p_6771(,|$))|((^|,)cat_p_6982(,|$))|((^|,)cat_c_2(,|$))|((^|,)cat_c_7(,|$))|((^|,)cat_c_p_7(,|$))"}} []
[2023-02-07T10:44:07.558501+00:00] main.DEBUG: cache_invalidate:  {"method":"GET","url":"http:/","invalidateInfo":{"tags":["cat_p_6771","cat_p_6982","cat_c_2","cat_c_7","cat_c_p_7"],"mode":"matchingAnyTag"}} []
[2023-02-07T10:44:07.558759+00:00] main.INFO: Cron Job indexer_update_all_views is successfully finished. Statistics: {"sum":2.2261419296265,"count":1,"realmem":10485760,"emalloc":22727128,"realmem_start":206045184,"emalloc_start":191820112} [] []
```

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by CptCharlesG** [comment]
@magento give me 2.4-develop instance

**Comment by CptCharlesG** [comment]
Well, I could not confirm it yet in the Magento Instance, at least what I see in page load waiting times.

After applying the same settings, creating a shipment makes no significant effect on loading time neither to the purchased product nor to other products. It's like there is no cache invalidation at all, not even for the purchased product.

First page load is 1.0-1.2s, and then all page loads ranges between 350-650ns whatever happens.

However it would be great to confirm cache invalidation in page headers or debug.log.

I will try to dig deeper in this issue. Maybe it is some 3rd party modules.

**Comment by engcom-November** [comment]
Hi @CptCharlesG ,
Thank you for reporting and collaboration. Verified the behavior on Magento 2.4-develop instance without any 3rd party modules installed / enabled and the issue is not reproducible. No issue observed in debug.log file. Also, I see cache is not invalidated after placing order or creating shipment and no difference observed observed in page loads. This is the Magento feature but not a bug. Magento 2 will clear correct caches on the frontend depending on cache tags. Each entity that impacts the cache will have it's own tags. Once a product is saved all of the pages on which that product appears will be cleared. No need to flush the cache Kindly refer the Magento [docs,](https://docs.magento.com/user-guide/v2.3/system/cache-management.html) [full page cache](https://docs.magento.com/user-guide/v2.3/system/cache-full-page.html) for further understanding of the cache management.
Thank you.

**Comment by engcom-November** [comment]
Hi @CptCharlesG ,
We are closing this issue as there has been no latest update on the same. Kindly reopen / create new issue if you are still facing any issues.
Thank you.

**Comment by mortenbirkelund** [comment]
@CptCharlesG I am experiencing the same issue. Did you dig deeper after your last comment?

**Comment by CptCharlesG** [comment]
@mortenbirkelund I have rewritten the whole updater with REST API calls, now everything works well and also it is faster.
I guess this is the preferred/better supported method for the future.

**Comment by ahmed-midori** [comment]
@CptCharlesG Hi, We are facing the exact same issue on 2.4.5-p2. Would be grateful if you could kindly share more information on how you resolved the issue?

Kind regards

**Comment by CptCharlesG** [comment]
@mortenbirkelund @ahmed-midori Do we have common addons (list above)?

@ahmed-midori :
So the original code was something like this (invalidates all cache):
`foreach(...) {
      $sourceItem = $this->sourceItemFactory->create();
      $sourceItem->setSourceCode('default');
      $sourceItem->setSku($sku);
      $sourceItem->setQuantity($qtyToSet);
      $sourceItem->setStatus(($qty > 0 ? 1 : 0));
      $this->sourceItemsSaveInterface->execute([$sourceItem]);
}`

New code with REST API calls, also managing reservations:

`// Gettings reservations
$connection = $this->resource->getConnection();
$reservationTable = $this->resource->getTableName('inventory_reservation');
$reservations = array();
$reservationResult = $connection->fetchAll('SELECT `sku`, SUM(`quantity`) as `qty` FROM `'.$reservationTable.'` GROUP BY `sku` ');
if ($reservationResult && count($reservationResult) > 0) {
	foreach ($reservationResult as &$reservation) {
		$reservations[$reservation['sku']] = $reservation['qty'];
	}
}

$qty2json = array();

// Preparing the json code
foreach ($data as $d) {
	$qtyToSet = $d['qty'];
	if (array_key_exists($d['sku'], $reservations)) $qtyToSet -= $reservations[$d['sku']];

	$qty2json[] = ["sku" => $d['sku'], "source_code" => "default", "quantity" => $qtyToSet, "status" => ($d['qty'] > 0 ? 1 : 0)];
}

try {
	// Get login token
	$userData = array("username" => "updater-admin-username", "password" => "updater-admin-password");
	$ch = curl_init("https://mydomain.com/rest/default/V1/integration/admin/token");
	curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
	curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($userData));
	curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
	curl_setopt($ch, CURLOPT_HTTPHEADER, array("Content-Type: application/json", "Content-Lenght: " . strlen(json_encode($userData))));

	$RESTtoken = curl_exec($ch);

	// Set quantity
	$ch = curl_init("https://mydomain.com/rest/default/V1/inventory/source-items");
	curl_setopt($ch, CURLOPT_TIMEOUT, 240);
	curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
	curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode(["sourceItems" => $qty2json]));
	curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
	curl_setopt($ch, CURLOPT_HTTPHEADER, array("Content-Type: application/json", "Authorization: Bearer " . json_decode($RESTtoken)));

	$result = curl_exec($ch);
	
	// $result = json_decode($result, true);
	// echo "<pre>".var_dump($result)."</pre>";
	if (strlen($result) > 10) {
		$error .= $result;
	}
	
} catch (\Exception $e) {
	$error .= $e->getMessage();
}`

Similarly we update prices by `products/base-prices` and `products/tier-prices`.

Notes:
We divide each run for a couple of hundreds of skus for each call to not overload the system.
It might be a good practice to compare the updated database with the wanted values to make sure everything is ok (on TODO list).

Reference:
[https://developer.adobe.com/commerce/webapi/rest/inventory/manage-source-items/](url)
[https://developer.adobe.com/commerce/webapi/rest/modules/catalog/catalog-pricing/](url)

I hope it helps.

**Comment by ahmed-midori** [comment]
@CptCharlesG

Thanks for your response. Much appreciated.

Some of the Amasty modules we have in common as well as Magefan blog.

The problem we were facing that with any inventory stock change from ad min or an shipment created the entire cache was flushed (absolutely everything). In troubleshooting I tried to disable all modules followed by updating them however neither had a change to this issue.

Following the update of all modules I decided to update magento to 2.4.6 from 2..4.5-p2 today and run some tests. Updating from 2.4.5-p2 to 2.4.6 seems to greatly have resolved the issue. Now a shipment or inventory level change made in admin only invalidates the cache for the specific product and not everything so much much better.

I have 2 instances running right now to compare. 2.4.5-p2 and 2.4.6. In my opinion this seems to be a bug in 2.4.5-p1/p2.

Once again thank you  @CptCharlesG for your quick response and detailed solution that worked for you.

Kind Regards
